File: SPAssistiveTouch_OC/SPAssistiveTouch.m
//
//  SPAssistiveTouch.m
//  alert
//
//  Created by 康世朋 on 2017/9/18.
//  Copyright © 2017年 康世朋. All rights reserved.
//

#import "SPAssistiveTouch.h"

@interface SPAssistiveTouch ()

@property (nonatomic, strong) UIView *parentView;
@property (nonatomic, assign) CGRect initialFrame;
@property (nonatomic, strong) UIButton *assistivaButton;

@property (nonatomic, strong) NSTimer *timer;

@property (nonatomic, assign) CGFloat transX;
@property (nonatomic, assign) CGFloat transY;
@property (nonatomic, assign) CGFloat maxX;
@property (nonatomic, assign) CGFloat maxY;

@property (nonatomic, copy) void(^clickBlock)(SPAssistiveTouch *touch);

@end

@implementation SPAssistiveTouch

+ (instancetype)showOnView:(UIView *)view x:(CGFloat)x y:(CGFloat)y width:(CGFloat)width {
    SPAssistiveTouch *ass = [[SPAssistiveTouch alloc]init];
    [ass setupUIWithView:view x:x y:y width:width];
    return ass;
}

+ (instancetype)showOnView:(UIView *)view x:(CGFloat)x y:(CGFloat)y width:(CGFloat)width click:(void (^)(SPAssistiveTouch *))block {
    SPAssistiveTouch *ass = [[SPAssistiveTouch alloc]init];
    ass.clickBlock = block;
    [ass setupUIWithView:view x:x y:y width:width];
    return ass;
}

- (void)setupUIWithView:(UIView *)view x:(CGFloat)x y:(CGFloat)y width:(CGFloat)width {
    
    self.frame = CGRectMake(x, y, width, width);
    _parentView = view;
    _initialFrame = self.frame;
    _stopScreenEdge = YES;
    _hasNavigationBar = YES;
    _alphaForStop = 0.4;
    _autoChangeAlpha = YES;
    self.layer.cornerRadius = width/2;
    _assistivaButton = [UIButton buttonWithType:UIButtonTypeCustom];
    _assistivaButton.frame = CGRectMake(0, 0, width, width);
    _assistivaButton.layer.cornerRadius = width/2;
    _assistivaButton.layer.masksToBounds = YES;
    _assistivaButton.backgroundColor = [UIColor blackColor];
    self.alpha = _alphaForStop;
    _assistivaButton.alpha = _alphaForStop;
    [_assistivaButton addTarget:self action:@selector(buttonClick) forControlEvents:UIControlEventTouchUpInside];
    [self addSubview:_assistivaButton];
    [view addSubview:self];
    UIPanGestureRecognizer *panGesture = [[UIPanGestureRecognizer alloc] initWithTarget:self action:@selector(pan:)];
    [self addGestureRecognizer:panGesture];
}

- (void)pan:(UIPanGestureRecognizer *)panGesture {
    [_timer invalidate];
    self.alpha = 1;
    _assistivaButton.alpha = 1;
    _transX = [panGesture translationInView:self].x;
    _transY = [panGesture translationInView:self].y;
    panGesture.view.center = CGPointMake(panGesture.view.center.x + _transX, panGesture.view.center.y + _transY);
    [panGesture setTranslation:CGPointZero inView:self];
    
    if (panGesture.state == UIGestureRecognizerStateEnded) {
        if (_autoChangeAlpha) {
           [self addTimer];
        }else {
            [_timer invalidate];
            _timer = nil;
        }
        
        if (self.frame.origin.y > _parentView.bounds.size.height - _initialFrame.size.height - 3) {
            _maxY = _parentView.bounds.size.height - _initialFrame.size.height - 3;
            [self resetFrameX:self.frame.origin.x y:_maxY w:self.frame.size.width h:self.frame.size.height];
        }
        if (_hasNavigationBar && self.frame.origin.y <= 67) {
            [self resetFrameX:self.frame.origin.x y:67 w:self.frame.size.width h:self.frame.size.height];
        }else if (!_hasNavigationBar && self.frame.origin.y <= 3) {
            [self resetFrameX:self.frame.origin.x y:3 w:self.frame.size.width h:self.frame.size.height];
        }
        if (self.frame.origin.x <= 3) {
            [self resetFrameX:3 y:self.frame.origin.y w:self.frame.size.width h:self.frame.size.height];
        }
        if (self.frame.origin.x > _parentView.bounds.size.width - _initialFrame.size.width - 3) {
            _maxX = _parentView.bounds.size.width - _initialFrame.size.width - 3;
            [self resetFrameX:_maxX y:self.frame.origin.y w:self.frame.size.width h:self.frame.size.height];
        }
        
        if (_stopScreenEdge && self.frame.origin.y < _parentView.frame.size.height - self.frame.size.height - 40 && self.frame.origin.y > 100) {
            if (self.frame.origin.x < _parentView.frame.size.width/2 - self.frame.size.width) {
                [self resetFrameX:3 y:self.frame.origin.y w:self.frame.size.width h:self.frame.size.height];
            }else {
                [self resetFrameX:_parentView.frame.size.width-3-self.frame.size.width y:self.frame.origin.y w:self.frame.size.width h:self.frame.size.height];
                
            }
        }else if (_stopScreenEdge && self.frame.origin.y > _parentView.frame.size.height - self.frame.size.height - 40) {
            _maxY = self.parentView.bounds.size.height - _initialFrame.size.height - 3;
            [self resetFrameX:self.frame.origin.x y:_maxY w:self.frame.size.width h:self.frame.size.height];
        }
        
        if (_stopScreenEdge && _hasNavigationBar && self.frame.origin.y < 100) {
            [self resetFrameX:self.frame.origin.x y:67 w:self.frame.size.width h:self.frame.size.height];
        }else if (_stopScreenEdge && !_hasNavigationBar && self.frame.origin.y < 100) {
            [self resetFrameX:self.frame.origin.x y:3 w:self.frame.size.width h:self.frame.size.height];
        }
    }
}

- (void)setBackColor:(UIColor *)backColor {
    _backColor = backColor;
    self.backgroundColor = backColor;
    _assistivaButton.backgroundColor = backColor;
}

- (void)setCornerRadius:(CGFloat)cornerRadius {
    _cornerRadius = cornerRadius;
    self.layer.cornerRadius = cornerRadius;
    _assistivaButton.layer.cornerRadius = cornerRadius;
}

- (void)setNormalImage:(UIImage *)normalImage {
    _normalImage = normalImage;
    [_assistivaButton setBackgroundImage:normalImage forState:UIControlStateNormal];
}

- (void)setLightImage:(UIImage *)lightImage {
    _lightImage = lightImage;
    [_assistivaButton setBackgroundImage:lightImage forState:UIControlStateHighlighted];
}

- (void)setAlphaForStop:(CGFloat)alphaForStop {
    _alphaForStop = alphaForStop;
    self.alpha = alphaForStop;
    _assistivaButton.alpha = alphaForStop;
}

- (void)setAutoChangeAlpha:(BOOL)autoChangeAlpha {
    _autoChangeAlpha = autoChangeAlpha;
    if (_autoChangeAlpha) {
        [self addTimer];
    }else {
        [_timer invalidate];
        _timer = nil;
    }
}

- (void)resetFrameX:(CGFloat)x y:(CGFloat)y w:(CGFloat)width h:(CGFloat)height {
    [UIView animateWithDuration:0.5 animations:^{
        self.frame = CGRectMake(x, y, width, height);
    }];
}

- (void)addTimer {
    [_timer invalidate];
    _timer = nil;
    _timer = [NSTimer scheduledTimerWithTimeInterval:2.0 target:self selector:@selector(changeAlpha) userInfo:nil repeats:YES];
    [[NSRunLoop mainRunLoop]addTimer:_timer forMode:NSRunLoopCommonModes];
}

- (void)buttonClick {
    if (_autoChangeAlpha) {
        [self addTimer];
    }else {
        [_timer invalidate];
        _timer = nil;
    }
    self.alpha = 1;
    _assistivaButton.alpha = 1;
    !_clickBlock? :_clickBlock(self);
}

- (void)changeAlpha {
    [UIView animateWithDuration:0.7 animations:^{
        self.alpha = self.alphaForStop;
        self.assistivaButton.alpha = self.alphaForStop;
    }];

}

/*
// Only override drawRect: if you perform custom drawing.
// An empty implementation adversely affects performance during animation.
- (void)drawRect:(CGRect)rect {
    // Drawing code
}
*/

@end

File: SPAssistiveTouch/SPAssistiveTouch.swift
//
//  SPAssistiveTouch.swift
//  Pom91
//
//  Created by 康世朋 on 16/8/4.
//  Copyright © 2016年 SP. All rights reserved.
//

import UIKit

class SPAssistiveTouch: UIView {
    
    var backColor = UIColor.gray {
        didSet {
            backgroundColor = backColor
            assistivaButton.backgroundColor = backColor
        }
    }
    
    
    var cornerRadius: CGFloat = 10.0 {
        didSet {
            layer.cornerRadius = cornerRadius
            assistivaButton.layer.cornerRadius = cornerRadius
        }
    }
    
    
    var normalImage: UIImage? {
        didSet{
           assistivaButton.setBackgroundImage(normalImage, for: UIControlState())
            assistivaButton.layer.cornerRadius = initialFrame.size.width/2
        }
    }
    
    var lightImage: UIImage? {
        didSet{
           assistivaButton.setBackgroundImage(lightImage, for: .highlighted)
            assistivaButton.layer.cornerRadius = initialFrame.size.width/2
        }
    }
    
    var alphaForlight: CGFloat = 1 {
        didSet {
            alpha = alphaForlight
            assistivaButton.alpha = alphaForlight
        }
    }
    
    var autoChangeAlpha = true {
        didSet {
            if autoChangeAlpha {
                addTimer()
            }else {
                timer.invalidate()
            }
        }
    }
    
    
    var alphaForStop: CGFloat = 0.5
    var stopScreenEdge = true
    var hasNavigationBar = true
    
    
    fileprivate var assistivaButton: UIButton! {
        didSet{
        }
    }
    fileprivate var timer = Timer()
    fileprivate var initialFrame: CGRect!
    fileprivate var parentView: UIView!
    func showInView(showOnview viw: UIView, X: CGFloat, Y: CGFloat, width: CGFloat) {
        setupUI(viw, X: X, Y: Y, width: width)

    }
    
    func add_target(target: AnyObject?, action: Selector) {
        assistivaButton.addTarget(target, action: action, for: .touchUpInside)
    }
    
    init(showOnView viw: UIView, X: CGFloat, Y: CGFloat, width: CGFloat) {
        super.init(frame: CGRect.zero)
        setupUI(viw, X: X, Y: Y, width: width)
        
    }
    
    init() {
        super.init(frame: CGRect.zero)
    }
    
    fileprivate func setupUI(_ viw: UIView, X: CGFloat, Y: CGFloat, width: CGFloat) {
        frame = CGRect(x: X, y: Y, width: width, height: width)
        parentView = viw
        initialFrame = frame
        layer.cornerRadius = width/2
        assistivaButton = UIButton(type: .custom)
        assistivaButton.frame = CGRect(x: 0, y: 0, width: width, height: width)
        assistivaButton.layer.cornerRadius = width/2
        assistivaButton.layer.masksToBounds = true
        assistivaButton.backgroundColor = backColor
        alpha = alphaForStop
        assistivaButton.alpha = alphaForStop
        addSubview(assistivaButton)
        viw.addSubview(self)
        addGestureRecognizer(UIPanGestureRecognizer(target: self, action: #selector(pan(_:))))
        
        addTimer()
    }
    
    fileprivate func addTimer() {
        timer = Timer(timeInterval: 3, target: self, selector: #selector(changeAlpha), userInfo: nil, repeats: true)
        RunLoop.main.add(timer, forMode: RunLoopMode.commonModes)
    }
    
    @objc fileprivate func changeAlpha() {
        UIView.animate(withDuration: 1, animations: { 
            self.alpha = self.alphaForStop
            self.assistivaButton.alpha = self.alphaForStop
        }) 
    }
    
    required init?(coder aDecoder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }
    

    @objc fileprivate func pan(_ panGesture: UIPanGestureRecognizer) {
        timer.invalidate()
        alpha = 1
        assistivaButton.alpha = 1
        let _transX: CGFloat = panGesture.translation(in: self).x
        let _transY: CGFloat = panGesture.translation(in: self).y
        var _maxY: CGFloat
        var _maxX: CGFloat
        
        transform = transform.translatedBy(x: _transX, y: _transY)
        panGesture.setTranslation(CGPoint.zero, in: self)
        
        if panGesture.state == .ended {
            addTimer()
            if frame.origin.y > parentView.bounds.height - initialFrame.size.height - 3 {
                 _maxY = parentView.bounds.height - initialFrame.size.height - 3
                
                resetFrame(x: self.frame.origin.x, y: _maxY, w: self.frame.size.width, h: self.frame.size.height)
            }
            if hasNavigationBar && frame.origin.y <= 67 {
        
                resetFrame(x: self.frame.origin.x, y: 67, w: self.frame.size.width, h: self.frame.size.height)
            }else if !hasNavigationBar && frame.origin.y <= 3 {
                resetFrame(x: self.frame.origin.x, y: 3, w: self.frame.size.width, h: self.frame.size.height)
            }
            if frame.origin.x <= 3 {
                resetFrame(x: 3, y: self.frame.origin.y, w: self.frame.size.width, h: self.frame.size.height)
            }
            if frame.origin.x > parentView.bounds.width - initialFrame.size.width - 3 {
                _maxX = parentView.bounds.width - initialFrame.size.width - 3
                resetFrame(x: _maxX, y: self.frame.origin.y, w: self.frame.size.width, h: self.frame.size.height)
            }
        
            if stopScreenEdge && frame.origin.y < parentView.frame.size.height - frame.size.height - 40 && frame.origin.y > 100{
                if frame.origin.x < parentView.frame.size.width/2 - frame.size.width {
                    resetFrame(x: 3, y: self.frame.origin.y, w: self.frame.size.width, h: self.frame.size.height)
                }else {
                    resetFrame(x: self.parentView.frame.size.width - 3 - self.frame.size.width, y: self.frame.origin.y, w: self.frame.size.width, h: self.frame.size.height)
                }
            }else if stopScreenEdge && frame.origin.y > parentView.frame.size.height - frame.size.height - 40 {
                _maxY = self.parentView.bounds.height - initialFrame.size.height - 3

                resetFrame(x: self.frame.origin.x, y: _maxY, w: self.frame.size.width, h: self.frame.size.height)
            }
            
            if stopScreenEdge && hasNavigationBar && frame.origin.y < 100 {
                resetFrame(x: self.frame.origin.x, y: 67, w: self.frame.size.width, h: self.frame.size.height)
            }else if stopScreenEdge && !hasNavigationBar && frame.origin.y < 100 {
                resetFrame(x: self.frame.origin.x, y: 3, w: self.frame.size.width, h: self.frame.size.height)
            }
        }
        
    }
    
    fileprivate func resetFrame(x: CGFloat, y: CGFloat, w:CGFloat, h: CGFloat) {
        UIView.animate(withDuration: 0.5, animations: {
            self.frame = CGRect(x: x, y: y, width: w, height: h)
        })
    }
    
    /*
    // Only override drawRect: if you perform custom drawing.
    // An empty implementation adversely affects performance during animation.
    override func drawRect(rect: CGRect) {
        // Drawing code
    }
    */

}
